libs/common/Networking/TcpNetworkHandlerBase.cs

```diff
@@ -135,31 +135,58 @@ void Dispose(SocketAsyncEventArgs e)
         }
 
         void RecvEventArg_Completed(object sender, SocketAsyncEventArgs e)
+        {
+            if (e.BytesTransferred == 0 || e.SocketError != SocketError.Success || serverHook.Disposed)
+            {
+                // No more things to receive
+                Dispose(e);
+                return;
+            }
+            var networkBufferProcessingTask = OnNetworkReceiveAsync(e.BytesTransferred);
+            if (networkBufferProcessingTask.IsCompletedSuccessfully)
+            {
+                NetworkBufferProcessed(networkBufferProcessingTask, sender, e);
+                return;
+            }
+            networkBufferProcessingTask.ContinueWith((res) =>
+            {
+                NetworkBufferProcessed(res, sender, e);
+            });
+        }
+
+        void NetworkBufferProcessed(Task result, object sender, SocketAsyncEventArgs e)
         {
             try
             {
-                do
+                if (result.IsCompletedSuccessfully)
                 {
-                    if (e.BytesTransferred == 0 || e.SocketError != SocketError.Success || serverHook.Disposed)
+                    e.SetBuffer(networkReceiveBuffer, networkBytesRead, networkReceiveBuffer.Length - networkBytesRead);
+                    if (!e.AcceptSocket.ReceiveAsync(e))
                     {
-                        // No more things to receive
-                        Dispose(e);
-                        break;
+                        RecvEventArg_Completed(sender, e);
                     }
-                    OnNetworkReceive(e.BytesTransferred);
-                    e.SetBuffer(networkReceiveBuffer, networkBytesRead, networkReceiveBuffer.Length - networkBytesRead);
-                } while (!e.AcceptSocket.ReceiveAsync(e));
+                }
+                else if (result.IsFaulted)
```

Comment:
don't we need another if for result.IsCanceled?

Reply:
Same as above